crates/hyperqueue/src/dashboard/data/tasks_timeline.rs

```diff
@@ -0,0 +1,61 @@
+use crate::server::event::events::MonitoringEventPayload;
+use crate::server::event::MonitoringEvent;
+use crate::WorkerId;
+use std::time::SystemTime;
+use tako::TaskId;
+
+#[derive(Clone)]
+pub struct TaskInfo {
+    pub worker_id: WorkerId,
+    pub task_id: TaskId,
+    pub start_time: SystemTime,
+    pub end_time: Option<SystemTime>,
+}
+
+impl TaskInfo {
+    pub fn set_end_time(&mut self, end_time: &SystemTime) {
+        self.end_time = Some(*end_time);
+    }
+}
+
+/// Stores information about the workers at different times
+#[derive(Default)]
+pub struct TasksTimeline {
+    task_timeline: Vec<TaskInfo>,
+}
+
+impl TasksTimeline {
+    /// Assumes that `events` are sorted by time.
+    pub fn handle_new_events(&mut self, events: &[MonitoringEvent]) {
+        for event in events {
+            match &event.payload {
+                MonitoringEventPayload::TaskStarted { task_id, worker_id } => {
+                    self.task_timeline.push(TaskInfo {
+                        worker_id: *worker_id,
+                        task_id: *task_id,
+                        start_time: event.time,
+                        end_time: None,
+                    });
+                }
+                MonitoringEventPayload::TaskFinished(finished_id) => {
+                    for info in &mut self.task_timeline {
```

Comment:
We will need to optimize a lot of things in the task timeline, but this could cause problems even in relatively small task graphs. Let's at least add a TODO comment to optimize this.